tests/backends/test_redis.py

```diff
@@ -153,6 +153,51 @@ async def test_redis(redisbackend):
         assert response.status_code == 200
 
 
+@pytest.mark.asyncio
+@pytest.mark.parametrize("redisbackend", [RedisBackend])
+async def test_multiple_(redisbackend):
+    await StrictRedis().flushdb()
+    rate_limit = RateLimitMiddleware(
+        hello_world,
+        auth_func,
+        redisbackend(),
+        {r"/multiple": [Rule(second=1, minute=3)]},
+    )
+    async with httpx.AsyncClient(
+        app=rate_limit, base_url="http://testserver"
+    ) as client:  # type: httpx.AsyncClient
+
+        # multiple 1/s and 3/min
+        # 1 3
+        response = await client.get(
+            "/multiple", headers={"user": "user", "group": "default"}
+        )
+        assert response.status_code == 200
+        # 1-1 3-1 = 0 2
+        response = await client.get(
+            "/multiple", headers={"user": "user", "group": "default"}
+        )
+        assert response.status_code == 429
+        await asyncio.sleep(1)
+        # 0+1 2+0 = 1 2
```

Comment:
> @euri10 Hey, after I re-read the code in detail and researched it, I found a problem. The two of us did not reach a consensus on the value deduction. In the original design, if the deduction fails, all values will not be deducted. Only when the deduction is successful, all values will be reduced by one.
> 
> I have submitted a new test and optimized the original code. Can you come and have a look when you are free?

yep that's a difference indeed between both implementations, in the `SlidingRedisBackend` a failed attempt counts against you everytime, I should have been clearer on this as it differs from the original implementation.

I'm not sure we can do something about as it is written as just one single lua script while the original `RedisBackend` does to round trips to to redis.

Reply:
Which one do you think is better? In any case, decrease it once, or decrease all counters by one only when it succeeds?
By the way, if you want, I would like to invite you to be a member of this warehouse so that you can directly make changes to this branch.

Reply:
> Which one do you think is better? In any case, decrease it once, or decrease all counters by one only when it succeeds?
> By the way, if you want, I would like to invite you to be a member of this warehouse so that you can directly make changes to this branch.

I think both are valid, as long as it matches users's expectations: a documentation note should therefore clear any ambiguity on this.

That said, I think that the original implementation is probably the most intuitive and the `SlidingRedisBackend` is less intuitive at first sight, but if you know about it, then it's ok.
The `SlidingRedisBackend` is more "punishing" : a "bad" user will see his counter decreased on any attempt.

Feel free to send an invite, it will be a pleasure :)

Reply:
I think your idea is great! Hahaha, I never thought about it before.
Then I will continue to optimize the original implementation. I think it should only need to execute a Lua code, just like the implementation you wrote.